manager/shard_strategy.go

```diff
@@ -0,0 +1,191 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flyteorg/flytepropeller/manager/config"
+	"github.com/flyteorg/flytepropeller/pkg/apis/flyteworkflow/v1alpha1"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+type ShardStrategy interface {
+	GetPodCount() int
+	UpdatePodSpec(pod *v1.PodSpec, podIndex int) error
```

Comment:
```suggestion
	UpdatePodSpec(pod *v1.PodSpec, podIndex int, maxPods int) error
```

What if we pass the maxPods as an input, this way, the shardStrategy is independent of the scale. This will allow us to - if we want in the future - create an auto-scaling flytepropeller?

Reply:
I guess you think - it will not work with per-namespace. but what if that is always `1` for per namespace and if we ever want to increase propellers per namespace we can?

Reply:
i'm having some trouble understanding this completely. as you mentioned in the case of hashing the shardStrategy can easily be independent of the scale, but it is very difficult to fit into the project and domain shard strategies - which are basically defined by scale.

my idea for supporting autoscaling just requires updating the ShardStrategy (so going the opposite direction of making them independent). for the case of HashShardStrategy if we increase / decrease the podCount it will change the hash value and consequently restart the necessary pods. it should be a really simple addition to the createPods loop that we currently have, we can add simple scale up and down functions onto the interface and in the loop something like:

```
if scaleUp {
    shardStrategy.ScaleUp()
} else if scaleDown {
    shardStrategy.ScaleDown()
}
```

as i mentioned, downstream the loop identifies the shardStrategy changes and will automatically update the running pods.

Reply:
ok